### ISSUE TITLE
Fine objects are sometimes smoothed away

### ISSUE THREAD

**Issue by e-k-m** [gengshan-y/high-res-stereo#22]
Thank you for making this code publicly available. I have been comparing the resulting
disparity maps using high-res-stereo with the pretrained model, with results using
OpenCV Semi Global Block Matcher using restrictive parameters for precise but
not dense disparity maps. In addition images of size 2736 x 1542 (w/h) are used as input for
the OpenCV SGMB matcher, while images of size 2736 x 1542 (w/h)  are also used as input
for the high-res-stereo matcher. Given this I have the following remarks and questions:

1. The results are generally very good concerning the density of estimated disparity
values and also precision.

2. In some cases fine features, such as curbs, the disparities are not correctly estimated. In detail, disparities are smoothed, do not have the value discontinuities they should actually have to represent the curbs. Yet, this is not always the case, it tends to be for homogeneous scenes  (e.g. overland scenes, and not for city scenes). 

To fix the issue 2. I have fine tuned the the existing model, using max-disp 384 and a extensive dataset (40 000 images) generated using  OpenCV SGMB matcher. In this dataset fine features are represented. Yet, this did not really help. I would now, like to train a fresh model, using larger input images and max-disp 768. According to my understanding all images are resized to 512 x 768 (https://github.com/gengshan-y/high-res-stereo/blob/master/dataloader/MiddleburyLoader.py#L62). What needs to be done to train with input images of larger size, in order to get usable / good results concerning fine features? Or in other words, what are possible solutions to get usable / good result for fine features?

Comparison images can be found here: https://varstuff.s3-eu-west-1.amazonaws.com/highresstereo/viewer.html with the right images == high-res-stereo results.

**Comment by gengshan-y** [comment]
Hi, thanks for the detailed description. I think I understand your goal.

> According to my understanding all images are resized to 512 x 768 

Images are not resized to 512x768. They are resized according to `rand_scale` and then cropped to 512x768.

Given that, here's what I would check to train a model using pseudo-ground-truth (SGBM),
- construct a dataloader for the new dataset and change the `rand_scale` to cover the designed scaling range (perhaps [0.25,1.2] , where 1.2 is the highest-possible scaling factor)
- make sure invalid disparities are set to values smaller than zero
- Combine the new dataset with the existing ones (Sceneflow, Middlebury, KITTI, etc.), also making sure all of them have similar number of images. Given the SGBM-ground-truth is sparser than others, you could multiply the dataset by a larger[ factor ](https://github.com/gengshan-y/high-res-stereo/blob/aae0b9b86c4ab007f83ed0f583f9ed7ff4b032ea/train.py#L86)

**Comment by e-k-m** [comment]
@gengshan-y thank you so much for your feedback. I am just implementing the changes. One little question, "make sure invalid disparities are set to values smaller than zero", do you consider all SGBM depth map values for which no depth map value exists (== 0) to be invalid disparities? For KITTI this seems to actually be the same case, but I can find the code part where such values are set < 0. Thanks and have a great day.

**Comment by gengshan-y** [comment]
The invalid pixels are masked [here](https://github.com/gengshan-y/high-res-stereo/blob/aae0b9b86c4ab007f83ed0f583f9ed7ff4b032ea/train.py#L105). You could set the invalid disparity values to 0 in the dataloader (or when creating the pesudo-GT dataset).

**Comment by e-k-m** [comment]
@gengshan-y super thanks again.
